### ISSUE TITLE
Inconsistent template behavior (bind vs. mixin)

### ISSUE THREAD

**Issue by gmpreussner** [nim-lang/Nim#3542]
The following setup works:

a.nim
```nim
template foo*() = discard
```

b.nim
```nim
template bar*() = foo()
```

c.nim
```nim
import a, b

foo()
```

However, if replace the template `a.foo` with an overload of an existing template, such as `system.pop`, the code no longer compiles, because the compiler only finds `system.pop`:

a.nim:
```nim
template pop*() = discard
```

b.nim
```nim
template foobar*() = pop()
```

c.nim
```nim
import a, b

foobar()
```

I find this behavior somewhat confusing, if not inconsistent. I don't know if this is desired behavior for bind vs. mixin template semantix, or whether it is a bug.

**Comment by gmpreussner** [comment]
Workaround: 

*b.nim*
```nim
template foobar*() =
  mixin pop
  pop()
```

**Comment by Araq** [comment]
The compiler and the spec agree. In your first example 'foo' is unbound identifier because system doesn't have 'foo', so it is put into a 'mixin' context. In your second example 'pop' is not unbound, but it is not overloaded either, so it is put into a 'bind' context. The manual needs to be updated with an example like this, but apart from that it works as indented.

**Comment by gmpreussner** [comment]
I feel that this is problematic though, because module *b.nim* cannot possibly know all existing symbols in advance. So it either needs to always use *mixin* for its template usages (which is cumbersome), or it may inadvertently break in the future if *c.nim* suddenly decided to import some other module that conflicts.

In other words: module *b.nim* can break by modifying unrelated module *c.nim*

**Comment by gmpreussner** [comment]
*a.nim*
```nim
template blah*() = discard
```

*b.nim*
```nim
template foobar*() = blah()
```

*c.nim*
```nim
import a, b

foobar()
```

*d.nim*
```nim
proc blah*() = discard
```

This will compile, but if we add ```import d``` to *c.nim*, then *b.nim* will fail to compile.

**Comment by Araq** [comment]
Yup. It's really bad, but I know of no alternative solution that is as convenient in practice. And also in practice it hasn't come up.

**Comment by Araq** [comment]
> This will compile, but if we add import d to c.nim, then b.nim will fail to compile.

This is a different topic though, IMO. The compiler is right and then it becomes ambiguous.

**Comment by gmpreussner** [comment]
Well, it came up for me today while writing a library of reusable templates :)

Can't the compiler know that both *template foo* and *proc foo* exist, and then pick the best match?

**Comment by Araq** [comment]
> Can't the compiler know that both template foo and proc foo exist, and then pick the best match?

But that is exactly what it does?

**Comment by gmpreussner** [comment]
No, it doesn't seem to find the mixin template if a symbol exists:

```
but expected one of: 
system.pop(s: var seq[T])
```
